raft/src/test/java/org/apache/kafka/raft/ProspectiveStateTest.java

```diff
@@ -0,0 +1,474 @@
+    public void testThreeNodeQuorumVoteRejected(boolean withDirectoryId) {
+        ReplicaKey node1 = replicaKey(1, withDirectoryId);
+        ReplicaKey node2 = replicaKey(2, withDirectoryId);
+        ProspectiveState state = newProspectiveState(
+            voterSetWithLocal(Stream.of(node1, node2), withDirectoryId)
+        );
+        assertFalse(state.epochElection().isVoteGranted());
+        assertFalse(state.epochElection().isVoteRejected());
+        assertEquals(Set.of(node1, node2), state.epochElection().unrecordedVoters());
+        assertTrue(state.recordRejectedVote(node1.id()));
+        assertEquals(Collections.singleton(node2), state.epochElection().unrecordedVoters());
+        assertFalse(state.epochElection().isVoteGranted());
+        assertFalse(state.epochElection().isVoteRejected());
+        assertTrue(state.recordRejectedVote(node2.id()));
+        assertEquals(Collections.emptySet(), state.epochElection().unrecordedVoters());
+        assertFalse(state.epochElection().isVoteGranted());
+        assertTrue(state.epochElection().isVoteRejected());
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testCanChangePreVote(boolean withDirectoryId) {
+        int voter1 = 1;
+        int voter2 = 2;
+        ProspectiveState state = newProspectiveState(
+            voterSetWithLocal(IntStream.of(voter1, voter2), withDirectoryId)
+        );
+        assertTrue(state.recordGrantedVote(voter1));
+        assertTrue(state.epochElection().isVoteGranted());
+        assertFalse(state.recordRejectedVote(voter1));
+        assertFalse(state.epochElection().isVoteGranted());
+
+        assertTrue(state.recordRejectedVote(voter2));
+        assertTrue(state.epochElection().isVoteRejected());
+        assertFalse(state.recordGrantedVote(voter2));
+        assertFalse(state.epochElection().isVoteRejected());
+    }
+
+
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testCannotGrantOrRejectNonVoters(boolean withDirectoryId) {
+        int nonVoterId = 1;
+        ProspectiveState state = newProspectiveState(voterSetWithLocal(IntStream.empty(), withDirectoryId));
+        assertThrows(IllegalArgumentException.class, () -> state.recordGrantedVote(nonVoterId));
+        assertThrows(IllegalArgumentException.class, () -> state.recordRejectedVote(nonVoterId));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = { true, false })
+    public void testIdempotentGrant(boolean withDirectoryId) {
+        int otherNodeId = 1;
+        ProspectiveState state = newProspectiveState(
+            voterSetWithLocal(IntStream.of(otherNodeId), withDirectoryId)
+        );
+        assertTrue(state.recordGrantedVote(otherNodeId));
+        assertFalse(state.recordGrantedVote(otherNodeId));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testIdempotentReject(boolean withDirectoryId) {
```

Comment:
Not sure if idempotent is the correct description. I would just call this test: testConsecutiveReject.